pkg/kubelet/kuberuntime/kuberuntime_manager.go

```diff
@@ -604,6 +632,64 @@ func (m *kubeGenericRuntimeManager) computePodActions(pod *v1.Pod, podStatus *ku
 			// If the container failed the startup probe, we should kill it.
 			message = fmt.Sprintf("Container %s failed startup probe", container.Name)
 		} else {
+			if utilfeature.DefaultFeatureGate.Enabled(features.InPlacePodVerticalScaling) {
+				keepCount++
+				if container.Resources.Limits == nil || len(pod.Status.ContainerStatuses) == 0 {
+					continue
+				}
+				apiContainerStatus, exists := podutil.GetContainerStatus(pod.Status.ContainerStatuses, container.Name)
+				if !exists || apiContainerStatus.State.Running == nil ||
+					containerStatus.State != kubecontainer.ContainerStateRunning ||
+					containerStatus.ID.String() != apiContainerStatus.ContainerID ||
+					len(diff.ObjectDiff(container.Resources.Requests, container.ResourcesAllocated)) != 0 ||
+					len(diff.ObjectDiff(apiContainerStatus.Resources.Limits, container.Resources.Limits)) == 0 {
+					continue
+				}
+				resizePolicy := make(map[v1.ResourceName]v1.ContainerResizePolicy)
+				for _, pol := range container.ResizePolicy {
+					resizePolicy[pol.ResourceName] = pol.Policy
+				}
+				restartContainer := false
+				determineContainerResize := func(rName v1.ResourceName, specValue, statusValue int64) bool {
+					if specValue == statusValue {
+						return false
+					}
+					if resizePolicy[rName] == v1.RestartContainer {
+						restartContainer = true
+						return true
+					}
+					cUpdateInfo := containerToUpdateInfo{
+						specIndex:           idx,
+						apiContainerStatus:  &apiContainerStatus,
+						kubeContainerStatus: containerStatus,
+						restart:             &restartContainer,
+					}
+					switch {
+					case specValue > statusValue: // append
+						changes.ContainersToUpdate[rName] = append(changes.ContainersToUpdate[rName], cUpdateInfo)
+					case specValue < statusValue: // prepend
+						changes.ContainersToUpdate[rName] = append(changes.ContainersToUpdate[rName], containerToUpdateInfo{})
+						copy(changes.ContainersToUpdate[rName][1:], changes.ContainersToUpdate[rName])
+						changes.ContainersToUpdate[rName][0] = cUpdateInfo
+					}
+					return false
+				}
+				specLim := container.Resources.Limits
+				statusLim := apiContainerStatus.Resources.Limits
+				restartMem := determineContainerResize(v1.ResourceMemory, specLim.Memory().Value(), statusLim.Memory().Value())
+				restartCPU := determineContainerResize(v1.ResourceCPU, specLim.Cpu().MilliValue(), statusLim.Cpu().MilliValue())
+				if restartMem || restartCPU {
+					// resize policy requires this container to restart
+					changes.ContainersToKill[containerStatus.ID] = containerToKillInfo{
+						name:      containerStatus.Name,
+						container: &pod.Spec.Containers[idx],
+						message:   fmt.Sprintf("Container %s resize requires restart", container.Name),
+					}
+					changes.ContainersToRestart = append(changes.ContainersToRestart, idx)
```

Comment:
can we just use ContainersToStart instead of creating a new ContainersToRestart?  I would hope ContainersToStart would start the container with its new resource requests.

Reply:
If the resize results in net-increase of pod resources, we need to update the pod-level cgroup values before starting the container. (Step 8 in SyncPod). We also need to apply the changes to any other containers that are being resized without restart so that we don't end up in a situation where the container cgroup values ever exceed the new pod-level cgroup values. I found adding ContainersToRestart invoked at the very end out resize operation pipeline as the cleanest way to achieve this.

Reply:
Actually, it is possible to get rid of ContainersToRestart if we switched resize step 8 with step 7. Functionally it should work just the same, I think.

However, if we cannot distinguish start from restart, then we would do some amount of possibly throw-away work (ResourceConfigForPod, get-current-pod-cg limit for cpu, mem) when no pod-level cgroup update is needed. If we do this for every new single container, we may see a performance regression.

Thoughts?

Reply:
I generally prefer starting with the simpler implementation and optimizing only if needed.

Reply:
ok, i'll make this change. we can optimize if there is a perf-regression.

Reply:
fixed. added a flag to distinguish restart from new container starts, so we shouldn't see a regression.